include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -983,6 +957,302 @@ struct __gen_expand_count_mask
     _RangeTransform __rng_transform;
 };
 
+// Locates and returns the "intersection" of a diagonal on the balanced path, based on merge path coordinates.
+// It returns coordinates in each set of the intersection with a boolean representing if the diagonal is "starred",
+// meaning that the balanced path "intersection" point does not lie directly on the diagonal, but one step forward in
+// the second set.
+// Some diagonals must be "starred" to ensure that matching elements between rng1 and rng2 are processed in pairs
+// starting from the first of repeating value(s) in each range and a matched pair are not split between work-items.
+template <typename _Rng1, typename _Rng2, typename _Index, typename _Compare>
+auto
+__find_balanced_path_start_point(const _Rng1& __rng1, const _Rng2& __rng2, const _Index __merge_path_rng1,
+                                 const _Index __merge_path_rng2, _Compare __comp)
+{
+    // back up to balanced path divergence with a biased binary search
+    bool __star = false;
+    if (__merge_path_rng1 == 0 || __merge_path_rng2 == __rng2.size())
+    {
+        return std::make_tuple(__merge_path_rng1, __merge_path_rng2, false);
+    }
+
+    auto __ele_val = __rng1[__merge_path_rng1 - 1];
+
+    if (__comp(__ele_val, __rng2[__merge_path_rng2]))
+    {
+        // There is no chance that the balanced path differs from the merge path here, because the previous element of
+        // rng1 does not match the next element of rng2. We can just return the merge path.
+        return std::make_tuple(__merge_path_rng1, __merge_path_rng2, false);
+    }
+
+    // find first element of repeating sequence in the first set of the previous element
+    _Index __rng1_repeat_start = oneapi::dpl::__internal::__biased_lower_bound</*__last_bias=*/true>(
+        __rng1, _Index{0}, __merge_path_rng1, __ele_val, __comp);
+    // find first element of repeating sequence in the second set of the next element
+    _Index __rng2_repeat_start = oneapi::dpl::__internal::__biased_lower_bound</*__last_bias=*/true>(
+        __rng2, _Index{0}, __merge_path_rng2, __ele_val, __comp);
+
+    _Index __rng1_repeats = __merge_path_rng1 - __rng1_repeat_start;
+    _Index __rng2_repeats_bck = __merge_path_rng2 - __rng2_repeat_start;
+
+    if (__rng2_repeats_bck >= __rng1_repeats)
+    {
+        // If we have at least as many repeated elements in rng2, we end up back on merge path
+        return std::make_tuple(__merge_path_rng1, __merge_path_rng2, false);
+    }
+    _Index __total_repeats = __rng1_repeats + __rng2_repeats_bck;
+
+    // Calculate the number of "unmatched" repeats in the first set, add one and divide by two to round up for a
+    // possible star diagonal.
+    _Index __fwd_search_count = (__rng1_repeats - __rng2_repeats_bck + 1) / 2;
```

Comment:
```suggestion
    _Index __fwd_search_count = (__rng1_repeats - __rng2_repeats_bck + 1) / 2;
    _Index __fwd_search_count = oneapi::dpl::__internal::__dpl_ceiling_div(__rng1_repeats - __rng2_repeats_bck, 2);
```
is it more correct?

Reply:
Technically I guess this would work.  However, I like it better as is conceptually. Also, we are in no danger of overflowing our type, so we can do one less addition this way.

Reply:
ok